pilight: repeated exception was filling openhab.log

While openhab was running, I edited some items. Some minutes afterwards, I observed that the pilight binding threw constantly one exception (several thousand times per second!). The openhab.log grew over 4 GB in size. I only could stop openhab to stop the exception.

The last two lines before it all started were:

```
2015-12-07 22:08:35.418 [INFO ] [.b.p.internal.PilightConnector] - Updating pilight config
2015-12-07 22:08:35.419 [ERROR] [.b.p.internal.PilightConnector] - Error in pilight listener thread
java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:209) ~[na:1.8.0_66]
        at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[na:1.8.0_66]
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284) ~[na:1.8.0_66]
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326) ~[na:1.8.0_66]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178) ~[na:1.8.0_66]
        at java.io.InputStreamReader.read(InputStreamReader.java:184) ~[na:1.8.0_66]
        at java.io.BufferedReader.fill(BufferedReader.java:161) ~[na:1.8.0_66]
        at java.io.BufferedReader.readLine(BufferedReader.java:324) ~[na:1.8.0_66]
        at java.io.BufferedReader.readLine(BufferedReader.java:389) ~[na:1.8.0_66]
        at org.openhab.binding.pilight.internal.PilightConnector.run(PilightConnector.java:115) ~[na:na]
```

After those lines, the following exception was repeated forever (quite similar to exception before, but started at a different line in PilightConnector.java):

```
2015-12-07 22:08:35.420 [ERROR] [.b.p.internal.PilightConnector] - Error in pilight listener thread
java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:154) ~[na:1.8.0_66]
        at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[na:1.8.0_66]
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284) ~[na:1.8.0_66]
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326) ~[na:1.8.0_66]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178) ~[na:1.8.0_66]
        at java.io.InputStreamReader.read(InputStreamReader.java:184) ~[na:1.8.0_66]
        at java.io.BufferedReader.fill(BufferedReader.java:161) ~[na:1.8.0_66]
        at java.io.BufferedReader.readLine(BufferedReader.java:324) ~[na:1.8.0_66]
        at java.io.BufferedReader.readLine(BufferedReader.java:389) ~[na:1.8.0_66]
        at org.openhab.binding.pilight.internal.PilightConnector.run(PilightConnector.java:93) ~[na:na]
```

--- a/bundles/binding/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/PilightBinding.java
+++ b/bundles/binding/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/PilightBinding.java
@@ -63,6 +63,12 @@ public class PilightBinding extends AbstractBinding<PilightBindingProvider> impl
 	
 	private Map<String, PilightConnection> connections = new HashMap<String, PilightConnection>();
 	
+	@Override
+	public void deactivate() {
+		logger.debug("pilight binding deactivated");
+		closeConnections();
+	}
+	
 	/**
 	 * Processes a status update received from pilight and changes the state of the
 	 * corresponding openHAB item (if item config is found)
@@ -235,6 +241,8 @@ public class PilightBinding extends AbstractBinding<PilightBindingProvider> impl
 	 */
 	@Override
 	public void updated(Dictionary<String, ?> config) throws ConfigurationException {
+		closeConnections();
+		
 		if (config != null) {			
 			this.connections = parseConfig(config);
 			
@@ -243,6 +251,15 @@ public class PilightBinding extends AbstractBinding<PilightBindingProvider> impl
 					PilightConnection connection = entry.getValue();
 					startConnector(connection);
 				}
+			}
+		}
+	}
+
+	private void closeConnections() {
+		for (Entry<String, PilightConnection> entry : connections.entrySet()) {
+			PilightConnection connection = entry.getValue();
+			if (connection.getConnector() != null) {
+				connection.getConnector().close();
 			}
 		}
 	}

--- a/bundles/binding/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/PilightConnector.java
+++ b/bundles/binding/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/PilightConnector.java
@@ -24,7 +24,6 @@ import org.codehaus.jackson.JsonGenerationException;
 import org.codehaus.jackson.JsonGenerator.Feature;
 import org.codehaus.jackson.map.JsonMappingException;
 import org.codehaus.jackson.map.ObjectMapper;
-import org.codehaus.jackson.map.SerializationConfig;
 import org.codehaus.jackson.map.annotate.JsonSerialize.Inclusion;
 import org.openhab.binding.pilight.internal.communication.Action;
 import org.openhab.binding.pilight.internal.communication.Identification;
@@ -104,7 +103,6 @@ public class PilightConnector extends Thread {
 								logger.trace("Response success: " + response.isSuccess());
 							} else if (line.equals("1")) {
 								// pilight stopping
-								connection.getSocket().close();
 								throw new IOException("Connection to pilight lost");
 							} else {
 								logger.debug(line);
@@ -116,14 +114,17 @@ public class PilightConnector extends Thread {
 					}
 				}
 			} catch (IOException e) {
-				logger.error("Error in pilight listener thread", e);
+				logger.debug("Error in pilight listener thread", e);
 			}
 			
-			// empty line received (socket closed) or pilight stopped, try to reconnect
-			reconnect();
+			logger.info("Disconnected from pilight server at {}:{}", connection.getHostname(), connection.getPort());
+
+			if (running) {
+				// empty line received (socket closed) or pilight stopped but binding 
+				// is still running, try to reconnect
+				reconnect();
+			}
 		}
-				
-		cleanup();
 	}
 	
 	/**
@@ -148,6 +149,7 @@ public class PilightConnector extends Thread {
 	 */
 	public void close() {
 		running = false;
+		disconnect();
 	}
 
 	/**
@@ -190,16 +192,19 @@ public class PilightConnector extends Thread {
 		} 
 	}
 	
-	private void cleanup() {
-		try {
-			connection.getSocket().close();
-		} catch (IOException e) {
-			logger.error("Error while closing pilight socket", e);
+	private void disconnect() {
+		if (connection.getSocket() != null) {
+			try {
+				connection.getSocket().close();
+			} catch (IOException e) {
+				logger.debug("Error while closing pilight socket", e);
+			}
 		}
-		logger.info("Thread pilight listener stopped");
 	}	
 
 	private void reconnect() {
+		disconnect();
+		
 		int delay = 0;
 		
 		while (!isConnected()) {
@@ -224,13 +229,13 @@ public class PilightConnector extends Thread {
 					logger.info("Established connection to pilight server at {}:{}", connection.getHostname(), connection.getPort());
 					connection.setSocket(socket);
 				} else {
-					logger.error("pilight client not accepted: {}", response.getStatus());
+					logger.debug("pilight client not accepted: {}", response.getStatus());
 				}
 				
 			} catch (IOException e) {
-				logger.error(e.getMessage(), e);
+				logger.debug(e.getMessage(), e);
 			} catch (InterruptedException e) {
-				logger.error(e.getMessage(), e);
+				logger.debug(e.getMessage(), e);
 			}
 			
 			delay = RECONNECT_DELAY;
@@ -238,10 +243,14 @@ public class PilightConnector extends Thread {
 	}
 	
 	public void doUpdate(Action action) {
-		if (connection.getDelay() != null) {
-			delayedUpdateCall(action);
+		if (isConnected()) {	
+			if (connection.getDelay() != null) {
+				delayedUpdateCall(action);
+			} else {
+				doUpdateCall(action);
+			}
 		} else {
-			doUpdateCall(action);
+			logger.debug("Cannot send command, not connected to pilight");
 		}
 	}
 
@@ -256,7 +265,7 @@ public class PilightConnector extends Thread {
 			connection.setLastUpdate(new Date());
 			outputMapper.writeValue(connection.getSocket().getOutputStream(), action);
 		} catch (IOException e) {
-			logger.error("Error while sending update to pilight server", e);
+			logger.debug("Error while sending update to pilight server", e);
 		}	
 	}
 	
@@ -288,7 +297,7 @@ public class PilightConnector extends Thread {
 					try {
 						Thread.sleep(delay);
 					} catch (InterruptedException e) {
-						logger.error("Error while processing pilight throttling delay");
+						logger.debug("Error while processing pilight throttling delay");
 					}
 				}
 			}

